src/python/zquantum/core/wip/circuits/_circuit.py

```diff
@@ -55,20 +55,24 @@ def operations(self):
         return self._operations
 
     @property
-    def n_qubits(self):
+    def n_qubits(self) -> int:
         """Number of qubits in this circuit.
         Not every qubit has to be used by a gate.
         """
         return self._n_qubits
 
     @property
-    def free_symbols(self):
+    def free_symbols(self) -> List[sympy.Symbol]:
         """Set of all the sympy symbols used as params of gates in the circuit."""
-        return reduce(
-            set.union,
-            (operation.gate.free_symbols for operation in self._operations),
-            set(),
-        )
+        seen_symbols = set()
+        symbols_sequence = []
+        for operation in self._operations:
+            for symbol in operation.gate.free_symbols:
+                if symbol not in seen_symbols:
+                    seen_symbols.add(symbol)
+                    symbols_sequence.append(symbol)
```

Comment:
I also asked @dexter2206 , and this is for performance reasons.
It's much faster to check `in` relation for set, as it's being hashed, rather than for a list.